packages/rocket-batch-file-process-aws-infrastructure/src/file-splitter-lambda/index.ts

```diff
@@ -16,34 +18,55 @@ export const handler = async (event: S3Event): Promise<void> => {
     Key: key,
   }
   const data = await s3.getObject(params).promise()
+  const fileExtension = key.split('.').pop()
+
+  if (fileExtension != CSV_FILE && fileExtension != JSONL_FILE) {
+    throw Error(`[ROCKET#batch-file] The provided fileExtension ${fileExtension} is not supported.`)
+  }
 
   if (data.Body) {
     const content = data.Body.toString()
 
     let chunkContent: string[] = []
-    const lines = content.split('\n')
-    let itr = 0
+    const rows = content.split('\n')
+
+    if (rows.length == 0) {
+      throw Error('[ROCKET#batch-file] The input file is empty.')
+    }
+
+    if (rows.length == 1) {
+      if (fileExtension == CSV_FILE) {
+        throw Error('[ROCKET#batch-file] Looks like the input CSV file only contains the headers row.')
+      } else {
+        throw Error('[ROCKET#batch-file] Looks like the input jsonl file only contains the first metadata row.')
+      }
+    }
+
+    let currentRowNumber = 0
     let fileNumber = 1
+    const firstRow: string = rows.shift()!
 
-    while (itr < lines.length) {
+    while (currentRowNumber < rows.length) {
       if (chunkContent.length != 0 && chunkContent.length % Number(process.env.CHUNK_SIZE) == 0) {
         const partKey = `part-${fileNumber}-${key}`
+        chunkContent.unshift(firstRow)
 
         await saveFile(partKey, chunkContent)
         await saveEvent(partKey)
 
         fileNumber = fileNumber + 1
         chunkContent = []
       }
-      if (lines[itr]) {
-        chunkContent.push(lines[itr])
+      if (rows[currentRowNumber]) {
+        chunkContent.push(rows[currentRowNumber])
       }
 
-      itr = itr + 1
+      currentRowNumber = currentRowNumber + 1
```

Comment:
I believe this `while` loop can be separated in two parts, the processing and the side effects:

```ts
// Processing part (probably should be a function)

const isWellFormedChunk = (chunkContent: Array<unknown>) => 
  chunkContent.length > 0 && chunkContent.length % chunkSize
const chunkSize = Number(process.env.CHUNK_SIZE ?? throw new Error('CHUNK_SIZE env var not set'))
const firstRow = rows[0] ?? throw new Error('Rows to be processed arrived empty')
const contentParts = 
  rows
    .filter(isWellFormedChunk)
    .map(row => [`part-${fileNumber}-${key}`, [firstRow, ...chunkContent, ...row])

// Side effects
for (const [partKey, chunkContent] of contentParts) {
  await saveFile(partKey, chunkContent)
  await saveEvent(partKey)
}
```

Reply:
One fun thing to consider is: what happens if there's an exception in one row in the middle? (i.e: the lambda could crash if the file is too long and the execution timeout is exceeded). One possible solution is having an extra DynamoDB (Or redis) table that records when a specific row is processed, so, if the process needs to be restarted, it knows where to continue.

I also have some doubts about using Lambda for this kind of workflow, because you'll have a lot of problems like this. I think that it could make a lot of sense to build a proper "job/worker" rocket where job orders, job statuses and data are persisted, the worker processes can run indefinitely (For instance, in a Fargate instance), and they can be safely stopped, restarted or scaled anytime (They'll just look at the jobs queues for things to do when they're running)

Reply:
@javiertoledo Deciding if Lambda is the right fit for this use case is of the tasks we have in mind actually.

The reasons you mentioned are not the ones I was thinking of but they totally makes sense (job order, job status) we can implement this in lambda but let's see.

As of now there are 2 things that concerns me in regards of Lambda

- How much will it cost? Is it going to be more expensive than their current solution with pods? We have a script they use with dummy data for their tests (the schema is accurate) so we'll try a few things with that

- Is the timeout going to be an issue? One way to mitigate this issue is the first lambda we have to split the file. Of course this totally dependes on the value of `chunkSize` which as of now is in total control of the user. We need to find out how long our lambdas are taking with the max number of records they handle in a single file (1 Million) We can do a test for 10K and then multiply and see. I agree this is not a stable solution as it totally depends on the schema and the `chunkSize` so depending on the case we can always hit the 15 min timeout but getting an idea of their most extreme case would give us and idea of where we are standing. 

My current reasoning is: If it turns out to be more expensive for similar workloads then we need to consider already a different alternative because they are price sensitive as our hackers already told us. If it happens to be about same or cheaper let's then check the timeout issue and see from there.

Fargate seems definitely a good alternative.

What do you think about this? @javiertoledo